### PR TITLE
Add zaza test coverage for ceph-mon reweight actions

### PR DESCRIPTION
**Pull request by anonybodi**
This adds zaza functional tests for ceph-mon reweight, reweight-by-utilization, and crush reweight actions. The addition of these actions to the ceph-mon charm can be viewed here: https://review.opendev.org/#/c/665719/

### CODE REVIEW

File: zaza/openstack/charm_tests/ceph/tests.py [hunk 1]
```diff
@@ -710,3 +710,119 @@ def test_ceph_health(self):
...
+            unit_name=unit_name,
+            action_name='reweight-by-utilization',
+            action_params=reweight_params
+        )
+        self.assertEqual('completed', action_obj.status)
+        zaza_model.block_until_unit_wl_status(
+            unit_name,
+            'active'
+        )
+        logging.debug('OK')
```
Comment (ChrisMacNaughton) [inline zaza/openstack/charm_tests/ceph/tests.py]: This test has validated that the action didn't fail, but this test hasn't validated anything about the actual weight of things, which has been modified in a (theoretically) predictable way.

File: zaza/openstack/charm_tests/ceph/tests.py [hunk 2]
```diff
@@ -710,3 +710,119 @@ def test_ceph_health(self):
...
+
+        The reweight-by-utilization action execute.
+        """
+        logging.info('Checking reweight-by-utilization action...')
+        unit_name = 'ceph-mon/0'
+
+        zaza_model.block_until_unit_wl_status(
+            unit_name,
+            'active'
+        )
```
[the remark this answers is not in the document]
Comment (anonybodi) [reply zaza/openstack/charm_tests/ceph/tests.py]: I have removed these checks from my latest commit.

File: zaza/openstack/charm_tests/ceph/tests.py [hunk 3]
```diff
@@ -710,3 +710,119 @@ def test_ceph_health(self):
...
+
+    @classmethod
+    def setUpClass(cls):
+        """Run class setup for running tests."""
+        super(CephMonReweightActionsTest, cls).setUpClass()
+
+    def test_reweight_by_utilization_no_threshold(self):
+        """Test the reweight-by-utilization (no threshold) action.
+
+        The reweight-by-utilization action execute.
```
[the remark this answers is not in the document]
Comment (anonybodi) [reply zaza/openstack/charm_tests/ceph/tests.py]: I have updated the comments on my latest commit.

File: zaza/openstack/charm_tests/ceph/tests.py (hunk 1 above)
Comment (anonybodi) [reply zaza/openstack/charm_tests/ceph/tests.py]: I have added checks (using ceph osd tree) for the reweight and crush rewight actions.

**Comment by ajkavanagh** [comment]
The PR has merged conflicts. @ChrisMacNaughton do you think it's still required as part of the ceph testing?

**Comment by ChrisMacNaughton** [comment]
This would still be a good test to have @ajkavanagh

File: zaza/openstack/charm_tests/ceph/tests.py [hunk 4]
```diff
@@ -710,3 +710,118 @@ def test_ceph_health(self):
             zaza_model.run_on_leader("ceph-proxy", "sudo ceph health")["Code"],
             "0"
         )
+
+
+class CephMonReweightActionsTest(unittest.TestCase):
+    """Test ceph mo reweight actions."""
```
Comment (ChrisMacNaughton) [inline zaza/openstack/charm_tests/ceph/tests.py]: typo here, on mon

**Comment by ajkavanagh** [comment]
@anonybodi If you are still interested/able to move this PR forwards then we'd love to get it incorporated.  It's just got a minor merge conflict and then it can be merged.

**Comment by anonybodi** [comment]
The merge conflict has been resolved and the typo fixed.

File: zaza/openstack/charm_tests/ceph/tests.py [hunk 5]
```diff
@@ -22,8 +22,8 @@
     path
 )
 import tempfile
-
 import tenacity
+import json
```
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/ceph/tests.py]: This line is causing the test errors; it's a repeat of line 18.

**Comment by ajkavanagh** [comment]
@ChrisMacNaughton is this still needed for the Ceph tests?  Added stale so that it can be closed if not needed.

**Comment by freyes** [comment]
I'm closing this since this PR has been stale for over a year now, although the testing looks interesting.

@sabaini , I'm closing this PR, although the Ceph team may have interest in resurrecting this patch and get it proposed in a new PR.

**Pull request closed without merging**